src/stream/src/executor/backfill.rs

```diff
@@ -111,57 +121,123 @@ where
+        // As such we will wait for next barrier.
         let is_snapshot_empty: bool = {
-            let snapshot = Self::snapshot_read(&self.table, init_epoch, None, false);
+            let snapshot = Self::snapshot_read(&self.upstream_table, init_epoch, None, false);
             pin_mut!(snapshot);
             snapshot.try_next().await?.unwrap().is_none()
         };
-        let to_backfill = to_create_mv && !is_snapshot_empty;
 
-        if to_create_mv && is_snapshot_empty {
-            // Directly finish the progress as the snapshot is empty.
-            self.progress.finish(first_barrier.epoch.curr);
-        }
+        // Backfill not Finished: Need to Backfill if snapshot is not empty.
+        // Backfill is  Finished: Need to Backfill not needed.
+        let to_backfill = if !is_finished {
+            !is_snapshot_empty
+        } else {
+            false
+        };
+
+        // Current position of the upstream_table storage primary key.
+        // `None` means it starts from the beginning.
+        let mut current_pos: Option<OwnedRow> = None;
+
+        // Use these to persist state.
+        // They contain the backfill position,
+        // as well as the progress.
+        // However, they do not contain the vnode mapping at index 0.
+        // That is filled in when we flush the state table.
+        let mut current_state: Vec<Datum> = vec![None; state_len];
+        let mut old_state: Option<Vec<Datum>> = None;
 
         // The first barrier message should be propagated.
         yield Message::Barrier(first_barrier);
 
+        // If no need backfill, but state was still "unfinished" we need to finish it.
+        // So we just process the next barrier to finish progress,
+        // and forward other messages.
+        //
+        // Reason for persisting on second barrier rather than first:
+        // We can't update meta with progress as finished until state_table
+        // has been updated.
+        // We also can't update state_table in first epoch, since state_table
+        // expects to have been initialized in previous epoch.
+        if !is_finished && !to_backfill {
+            while let Some(Ok(msg)) = upstream.next().await {
+                match &msg {
+                    Message::Barrier(barrier) => {
+                        current_pos =
+                            Self::construct_initial_finished_state(pk_in_output_indices.len());
+                        Self::persist_state(
+                            barrier.epoch,
+                            &mut self.state_table,
+                            true,
+                            &current_pos,
+                            &mut old_state,
+                            &mut current_state,
+                        )
+                        .await?;
+                        self.progress.finish(barrier.epoch.curr);
+                        yield msg;
+                        break;
+                    }
```

Comment:
Please take note, instead updating progress at first barrier, I update it in the second barrier.
Reasons are mentioned in the comment. TLDR: state table currently can only be committed in second barrier onwards, so I also update progress to meta from second barrier onwards.

Reply:
This also means I should check the snapshot at this epoch instead of the first one.

Including this change in a bit.

Reply:
Is it the only reason why recovery tests failed before?